### ISSUE TITLE
New Feature - Show information on top of unknown cards

### ISSUE THREAD

**Issue by Zamiell** [Zamiell/keldon-hanabi#59]

**Comment by Zamiell** [comment]
Thanks for your work on this Libster; it's looking excellent so far.

@aliblong  Please document your progress and current TODO list, as there's a lot to do yet I think.

**Comment by aliblong** [comment]
eb98b0ad1121f1a16881746d4671c00bbe990bd2 is the first functional version of this update. Each card has a set of number pips and a set of suit pips which disappear as positive and negative information is gained on the card in order to show the possible identities of the card. When the true suit or rank is known, the suit or rank pips will disappear and the the card will be colored or numbered, respectively, as was done in the replay previously.

With 4bc77b7545d0f6f0a01d4669098ed6087d9d5100, I reintroduced `unknownRect`, which is a 0.4 opacity white layer that sits over certain cards in a replay. Now, it is used at points in the replay when the true suit or rank of a card is unknown, despite complete knowledge of the respective quantity in the present. In [this example](http://i.imgur.com/fNHIfHq.png), the b1 is faded but not the b3, since at this point in the replay the former has incomplete rank information but the latter is known as true b3. This is to mitigate visual clash between the card face art and the pips. However, it's not an ideal system, leading to my next point.

@rogers-j came up with an idea to use a pips-only system in these instances, in which pips corresponding to present knowledge are highlighted (or their counterparts faded). I expect this to be a more aesthetically pleasing solution, and on top of that, it will preserve negative present knowledge, unlike the current system described in the first paragraph. Implementation-wise, it should require tracking the complete positive and negative information on `learnedCards`, which should be simple to implement.

In addition, I would like to introduce the ability to see what knowledge one's teammates have at a given point in time, i.e. display exactly how the cards in their hand look from their seat. This could be done by using a key or button to toggle this view, or by mousing over individual cards, or both. I expect this change will be a bit more time-consuming to implement than the previous one.

In summary, TODO:

- Implement highlighted/faded pips system for replay knowledge

- Implement teammate view

**Comment by Zamiell** [comment]
I think that together, when deciding which system to use, we should place a somewhat high value in simplicity, even if it encroaches upon being able to see "all the information, all the time, all at once".

Having a system where pips are highlighted or faded... well I think it sounds like it is getting to the point where it might be getting too overly complex. Is it really necessary, for example, in your screenshot, to be able to see current + past knowledge of BOTH b1 and b3 at the same time?

It is my suspicion that we can gut a lot of complexity here. What if we can just make "knowledge at the time" the default, and then upon mouseover it reveals the present information.

Thoughts on this? @rogers-j @aliblong

**Comment by aliblong** [comment]
I think the easiest way to get an idea of whether or not it will be information overload is just to implement and test it. Then we can reconvene.

**Comment by rogers-j** [comment]
I think in both cases we allow this complexity because it's representing something complex in a useful way. By failing to address the issue we make the user do worse things with notes or with the UI.

This is only complexity during the replay for a card that has has additional meaningful information since the replay position (`replayPos`).

**Comment by rogers-j** [comment]
Completed as of 990e4e20c33
